src/arch/riscv/object/objecttype.c

```diff
@@ -52,6 +59,25 @@ deriveCap_ret_t Arch_deriveCap(cte_t *slot, cap_t cap)
 
 cap_t CONST Arch_updateCapData(bool_t preserve, word_t data, cap_t cap)
 {
+#ifdef CONFIG_ALLOW_SBI_CALLS
+    if (cap_get_capType(cap) == cap_sbi_cap) {
+        if (preserve) {
+            return cap_null_cap_new();
+        } else {
+            if (cap_sbi_cap_get_capSBIEIDBadge(cap)) {
+                if (cap_sbi_cap_get_capSBIFIDBadge(cap)) {
+                    return cap_null_cap_new();
+                } else {
+                    cap_t badged_cap = cap_sbi_cap_set_capSBIFIDBadge(cap, data);
+                    return cap_sbi_cap_set_capSBIFIDBadged(badged_cap, 1);
+                }
+            } else {
+                cap_t badged_cap = cap_sbi_cap_set_capSBIEIDBadge(cap, data);
+                return cap_sbi_cap_set_capSBIEIDBadged(badged_cap, 1);
```

Comment:
What about invalid badge values because they wouldn't fit? I suppose silently ignoring them makes sense, but in debug builds it probably triggers an assert. Is that what we want though? I guess the same happens for e.g. CNode guards.